### PR TITLE
fix: crash after nullptr dereference in rpcimpl

### LINKED ISSUE #6172: segfault main
### What is the issue?

from main github compiled today :

```
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `transmission-daemon --log-debug -f'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00005610af490ddd in tr_torrent::stats (this=this@entry=0x0) at /var/www/transmission/git/transmission/libtransmission/torrent.cc:1347
1347        auto const swarm_stats = this->swarm != nullptr ? tr_swarmGetStats(this->swarm) : tr_swarm_stats{};
[Current thread is 1 (Thread 0x7f0f949ff640 (LWP 72574))]

```

full bt:
```
(gdb) bt full
#0  0x00005610af490ddd in tr_torrent::stats (this=this@entry=0x0) at /var/www/transmission/git/transmission/libtransmission/torrent.cc:1347
        now_msec = <optimized out>
        now_sec = <optimized out>
        swarm_stats = <optimized out>
        activity = <optimized out>
        idle_seconds = <optimized out>
        stats = {errorString = 0x3f <error: Cannot access memory at address 0x3f>, sizeWhenDone = 816, leftUntilDone = 12, desiredAvailable = 257698037809, corruptEver = 1, uploadedEver = 0, downloadedEver = 0,
          haveValid = 511101108348, haveUnchecked = 395136991342, addedDate = 419397718378029505, doneDate = 0, startDate = -25720, activityDate = 768, editDate = 139704817156144, recheckProgress = 0, percentComplete = 0,
          metadataPercentComplete = -1.90584153e-28, percentDone = 4.55800351e-41, seedRatioPercentDone = 0, pieceUploadSpeed_KBps = 0, pieceDownloadSpeed_KBps = -4.7319997e-26, ratio = 4.55800351e-41, id = -1801485696,
          idleSecs = 94629660545000, secondsDownloading = 768, secondsSeeding = 139704894718640, queuePosition = 0, eta = 139704912243084, etaIdle = 139704894718640, activity = 2940468744, error = 22032, peersConnected = 1,
          peersFrom = {0, 0, 0, 44944, 37232, 32527, 0}, peersSendingToUs = 0, peersGettingFromUs = 0, webseedsSendingToUs = 0, finished = false, isStalled = false}
        piece_upload_speed_byps = <optimized out>
        piece_download_speed_byps = <optimized out>
        verify_progress = <optimized out>
        seed_ratio_bytes_left = <optimized out>
        seed_ratio_bytes_goal = <optimized out>
        seed_ratio_applies = <optimized out>
#1  0x00005610af4919ed in tr_torrentStat (tor=tor@entry=0x0) at /var/www/transmission/git/transmission/libtransmission/torrent.cc:1465
No locals.
#2  0x00005610af59361a in (anonymous namespace)::addTorrentInfo (tor=tor@entry=0x0, format=format@entry=(anonymous namespace)::TrFormat::Object, entry=entry@entry=0x7f0f9170a490, fields=fields@entry=0x7f0f9000ff40,
    field_count=field_count@entry=12) at /var/www/transmission/git/transmission/libtransmission/rpcimpl.cc:728
        st = <optimized out>
#3  0x00005610af593e2e in (anonymous namespace)::addTorrentInfo (field_count=<optimized out>, fields=<optimized out>, entry=0x7f0f9170a490, format=(anonymous namespace)::TrFormat::Object, tor=0x0)
    at /var/www/transmission/git/transmission/libtransmission/rpcimpl.cc:726
        st = <optimized out>
        i = <optimized out>
        child = <optimized out>
#4  (anonymous namespace)::torrentGet (session=<optimized out>, args_in=<optimized out>, args_out=<optimized out>) at /var/www/transmission/git/transmission/libtransmission/rpcimpl.cc:797
        tor = 0x0
        __for_range = std::vector of length 49, capacity 49 = {0x0, 0x7f0f9000e560, 0x7f0f90016db0, 0x7f0f90018360, 0x7f0f90019790, 0x7f0f9001b700, 0x7f0f9008b2c0, 0x7f0f90092ac0, 0x7f0f900aa2f0, 0x7f0f900afb60, 0x7f0f900b0a00,
          0x7f0f902abfb0, 0x7f0f902fb0b0, 0x7f0f9035d2e0, 0x7f0f9030b470, 0x7f0f90314a30, 0x7f0f9034efb0, 0x7f0f90350f50, 0x7f0f90317bd0, 0x7f0f903ffd70, 0x7f0f90406770, 0x7f0f90402020, 0x7f0f903f13f0, 0x7f0f903241e0,
          0x7f0f903a5480, 0x7f0f9040e480, 0x7f0f9040fd40, 0x7f0f9041f0f0, 0x7f0f90420400, 0x7f0f904240f0, 0x7f0f9041a430, 0x7f0f903b0850, 0x7f0f903b3f10, 0x7f0f90451770, 0x7f0f90592a40, 0x7f0f90597610, 0x7f0f905a1c20,
          0x7f0f90595740, 0x7f0f905bb6f0, 0x7f0f904dedf0, 0x7f0f90535890, 0x7f0f90537330, 0x7f0f9072fbf0, 0x7f0f907357c0, 0x7f0f9079ede0, 0x7f0f90806100, 0x7f0f907a3b30, 0x7f0f90ac0e10, 0x7f0f90014530}
        __for_begin = <optimized out>
        __for_end = <optimized out>
        n = <optimized out>
        keys = std::vector of length 12, capacity 12 = {189, 218, 124, 366, 250, 121, 127, 69, 345, 92, 93, 75}
        list = 0x7f0f91719818
        torrents = std::vector of length 49, capacity 49 = {0x0, 0x7f0f9000e560, 0x7f0f90016db0, 0x7f0f90018360, 0x7f0f90019790, 0x7f0f9001b700, 0x7f0f9008b2c0, 0x7f0f90092ac0, 0x7f0f900aa2f0, 0x7f0f900afb60, 0x7f0f900b0a00,
          0x7f0f902abfb0, 0x7f0f902fb0b0, 0x7f0f9035d2e0, 0x7f0f9030b470, 0x7f0f90314a30, 0x7f0f9034efb0, 0x7f0f90350f50, 0x7f0f90317bd0, 0x7f0f903ffd70, 0x7f0f90406770, 0x7f0f90402020, 0x7f0f903f13f0, 0x7f0f903241e0,
          0x7f0f903a5480, 0x7f0f9040e480, 0x7f0f9040fd40, 0x7f0f9041f0f0, 0x7f0f90420400, 0x7f0f904240f0, 0x7f0f9041a430, 0x7f0f903b0850, 0x7f0f903b3f10, 0x7f0f90451770, 0x7f0f90592a40, 0x7f0f90597610, 0x7f0f905a1c20,
          0x7f0f90595740, 0x7f0f905bb6f0, 0x7f0f904dedf0, 0x7f0f90535890, 0x7f0f90537330, 0x7f0f9072fbf0, 0x7f0f907357c0, 0x7f0f9079ede0, 0x7f0f90806100, 0x7f0f907a3b30, 0x7f0f90ac0e10, 0x7f0f90014530}
        sv = "downloadDir"
        format = (anonymous namespace)::TrFormat::Object
        fields = 0x7f0f90096b78
        errmsg = 0x0
#5  0x00005610af58b9d5 in tr_rpc_request_exec_json (session=0x5610afa27870, request=0x7f0f949f8510, callback=callback@entry=0x5610af57fc50 <(anonymous namespace)::rpc_response_func(tr_session*, tr_variant*, void*)>,
    callback_user_data=0x7f0f9092fbb0) at /var/www/transmission/git/transmission/libtransmission/rpcimpl.cc:2399
        response = {val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 6] = {{vec_ = std::vector of length 1, capacity 3 = {{first = 25, second = {
                    val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 6] = {{vec_ = std::vector of length 1, capacity 1 = {{first = 365, second = {
                              val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 5] containing std::vector of length 1, capacity 50 = {{
                                  val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 6] = {{vec_ = std::vector of length 0, capacity 12}}}}}}}}}}}}}}}
        args_out = <optimized out>
        lock = <optimized out>
        mutable_request = 0x7f0f949f8510
        args_in = 0x7f0f91719688
        result = 0x0
        sv = "torrent-get"
        method = 0x5610af6948e0 <(anonymous namespace)::Methods+448>
#6  0x00005610af586c8d in (anonymous namespace)::handle_rpc_from_json (json=..., server=0x5610afa31bf0, req=0x7f0f900057d0) at /usr/include/c++/11/optional:432
        otop = std::optional<tr_variant> = {[contained value] = {val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 6] = {{
                vec_ = std::vector of length 2, capacity 2 = {{first = 184, second = {val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 4] = {{
                          sv_ = "torrent-get", str_ = "torrent-get"}}}}, {first = 25, second = {val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 6] = {
                        {vec_ = std::vector of length 1, capacity 1 = {{first = 96, second = {
                                val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 5] containing std::vector of length 12, capacity 16 = {{
                                    val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 4] = {{sv_ = "name", str_ = "name"}}}, {
                                    val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 4] = {{sv_ = "peersSendingToUs", str_ = "peersSendingToUs"}}}, {
                                    val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 4] = {{sv_ = "haveValid", str_ = "haveValid"}}}, {
                                    val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 4] = {{sv_ = "totalSize", str_ = "totalSize"}}}, {
                                    val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 4] = {{sv_ = "rateDownload", str_ = "rateDownload"}}}, {
                                    val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 4] = {{sv_ = "hashString", str_ = "hashString"}}}, {
                                    val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 4] = {{sv_ = "id", str_ = "id"}}}, {
                                    val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 4] = {{sv_ = "doneDate", str_ = "doneDate"}}}, {
                                    val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 4] = {{sv_ = "status", str_ = "status"}}}, {
                                    val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 4] = {{sv_ = "error", str_ = "error"}}}, {
                                    val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 4] = {{sv_ = "errorString", str_ = "errorString"}}}, {
                                    val_ = std::variant<std::monostate, bool, long, double, tr_variant::StringHolder, std::vector<tr_variant>, tr_variant::Map> [index 4] = {{sv_ = "downloadDir", str_ = "downloadDir"}}}}}}}}}}}}}}}}


        otop = <optimized out>
#7  (anonymous namespace)::handle_rpc (server=0x5610afa31bf0, req=0x7f0f900057d0) at /var/www/transmission/git/transmission/libtransmission/rpc-server.cc:375
Python Exception <class 'gdb.error'>: value has been optimized out
        json =
#8  (anonymous namespace)::handle_request (req=0x7f0f900057d0, arg=0x5610afa31bf0) at /var/www/transmission/git/transmission/libtransmission/rpc-server.cc:578
        uri = <optimized out>
        location = <optimized out>
        HttpErrorUnauthorized = 401
        HttpErrorForbidden = 403
        server = 0x5610afa31bf0
#9  0x00007f0f963e6e9c in ?? () from /lib/x86_64-linux-gnu/libevent-2.1.so.7
No symbol table info available.
#10 0x00007f0f963c02dc in bufferevent_run_readcb_ () from /lib/x86_64-linux-gnu/libevent-2.1.so.7
No symbol table info available.
#11 0x00007f0f963c2e3c in ?? () from /lib/x86_64-linux-gnu/libevent-2.1.so.7
No symbol table info available.
#12 0x00007f0f963caf58 in ?? () from /lib/x86_64-linux-gnu/libevent-2.1.so.7
No symbol table info available.
#13 0x00007f0f963cc8a7 in event_base_loop () from /lib/x86_64-linux-gnu/libevent-2.1.so.7
No symbol table info available.
#14 0x00005610af5a1dc9 in tr_session_thread_impl::sessionThreadFunc (this=0x5610afa27340, evbase=0x5610afa27450) at /var/www/transmission/git/transmission/libtransmission/session-thread.cc:239
        ToggleLooping = <optimized out>
#15 0x00007f0f95adc253 in ?? () from /lib/x86_64-linux-gnu/libstdc++.so.6
No symbol table info available.
#16 0x00007f0f95694ac3 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
        ret = <optimized out>
        pd = <optimized out>
        out = <optimized out>
        unwind_buf = {cancel_jmp_buf = {{jmp_buf = {140724486804448, -1311545559836930373, 139704894748224, 2, 139704907941840, 140724486804800, 1428588294247927483, 1428589486063734459}, mask_was_saved = 0}}, priv = {pad = {0x0,
              0x0, 0x0, 0x0}, data = {prev = 0x0, cleanup = 0x0, canceltype = 0}}}
        not_first_call = <optimized out>
#17 0x00007f0f95726a40 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
```

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

git

### PR DESCRIPTION
**Pull request by nevack**
Follow up for #5604
Fixes #6172

### CODE REVIEW

**Comment by nevack** [comment]
@ckerr
Entries are not removed from `by_ids_`, but rather being nulled out
https://github.com/transmission/transmission/blob/main/libtransmission/torrents.cc#L90

Also there's one always nullptr entry in `by_ids_`
https://github.com/transmission/transmission/blob/main/libtransmission/torrents.h#L130 

Another possible fix would be putting null-check here:
https://github.com/transmission/transmission/blob/main/libtransmission/rpcimpl.cc#L797

But there's a lot of other `getTorrents` calls, and I don't really know, if other places are working with nullptr "holes" in this vector.

File: libtransmission/rpcimpl.cc [hunk 1]
```diff
@@ -134,7 +134,7 @@ auto getTorrents(tr_session* session, tr_variant* args)
                 std::begin(by_id),
```
Comment (1100101) [inline libtransmission/rpcimpl.cc:134]:
```suggestion
                std::begin(by_id)+1,
```
You could skip the first entry, as it'll always be `nullptr`

File: libtransmission/rpcimpl.cc [hunk 2]
```diff
@@ -148,7 +148,12 @@ auto getTorrents(tr_session* session, tr_variant* args)
     else // all of them
     {
         auto const& by_id = session->torrents().sorted_by_id();
-        torrents = std::vector<tr_torrent*>{ std::begin(by_id), std::end(by_id) };
+        torrents.reserve(std::size(by_id));
+        std::copy_if(
+            std::begin(by_id),
+            std::end(by_id),
+            std::back_inserter(torrents),
+            [](auto const* tor) { return tor != nullptr; });
```
Comment (1100101) [inline libtransmission/rpcimpl.cc:156]: This change may not be necessary, at least not if any caller of the function correctly checks each entry of the vector for `NULL`.
But, we could skip the first entry here as well:
```suggestion
        torrents = std::vector<tr_torrent*>{ std::begin(by_id)+1, std::end(by_id) };
```

Comment (nevack) [reply libtransmission/rpcimpl.cc:156]: There can be many `nullptr` values in between, not only the first one is null here.
Please read my comment https://github.com/transmission/transmission/pull/6177#issuecomment-1783956064

File: libtransmission/rpcimpl.cc (hunk 1 above)
Comment (nevack) [reply libtransmission/rpcimpl.cc:134]: It doesn't make any difference, tbh, as it won't be copied.

File: libtransmission/rpcimpl.cc (hunk 2 above)
Comment (1100101) [reply libtransmission/rpcimpl.cc:156]: I understand. But any method that access the returned list of `torrents` should check for `NULL` anyway, so an additional NULL check here might not be necessary.
Anyway, performance-wise it won't make much of a difference, I suppose?

Comment (nevack) [reply libtransmission/rpcimpl.cc:156]: It won't matter performance wise.
But I would better fix it in one place according to the current invariant - no nulls returned - rather then update all 10+ call sites with null checks.

Comment (ckerr) [reply libtransmission/rpcimpl.cc:156]: @1100101 is right that the first element will always be nullptr and so it _could_ be skipped, but IMO hardcoding a `+1` here is RPC making minor optimizations based on knowledge of how `tr_torrents` is coded. I could imagine this biting us if `tr_torrents` ever gets refactored.

**Review by ckerr — APPROVED** [review]
This is a nice catch and a nice fix.

Looking over the PR that this fixes, it looks to me like it also introduced some overlapping responsibilities between `tr_session::getAllTorrents()` and `tr_torrents::sorted_by_id()`.

I've created a new PR #6179 which keeps @nevack's authorship on the new test code and which adds the nullptr check in a way fixes the overlapping responsiblities between `tr_session` and `tr_torrents`.

@nevack WDYT?

**Comment by nevack** [comment]
> WDYT?

I think you can merge my fix PR as is, and then refactor the way you want.
It's not like I prefer my code over your code, not like this 😄 
Personally, I'm against in PRs that mix these two things, we need a stable fix, to be able to bisect problems in future.

But it's more like a preference, I won't mind if you decide to do it your way.
I just have less expertise in the code base to verify this refactor fixes the problem without introducing new one.

**Pull request merged**